File: web/src/engine/fen.ts
import {
  BOARD_SIZE,
  type CastlingRights,
  engineNotation,
  fromEngineNotation,
  fromStandardNotation,
  type Piece,
  type PieceKind,
  type Square,
} from "./types";
import { ChessGame, squaresEqual } from "./chessGame";
import { resolveUciInterpretations } from "./uci";

export function toFEN(game: ChessGame): string {
  const ranks: string[] = [];
  for (let row = 0; row < BOARD_SIZE; row++) {
    ranks.push(fenRank(game, row));
  }
  const placement = ranks.join("/");
  const side = game.activeColor === "white" ? "w" : "b";
  const castle = fenCastling(game);
  const ep = game.enPassantTarget ? engineNotation(game.enPassantTarget) : "-";
  return `${placement} ${side} ${castle} ${ep} ${game.halfmoveClock} ${game.fullmoveNumber}`;
}

/** Load a Fairy-Stockfish chessborder FEN (for tests and diagnostics). */
export function fromFEN(fen: string): ChessGame {
  const parts = fen.trim().split(/\s+/);
  if (parts.length < 4) throw new Error("Invalid FEN");

  const ranks = parts[0].split("/");
  if (ranks.length !== BOARD_SIZE) throw new Error("Invalid FEN board");

  const game = new ChessGame();
  for (let row = 0; row < BOARD_SIZE; row++) {
    parseFenRank(ranks[row] ?? "", row, game.board);
  }

  game.activeColor = parts[1] === "b" ? "black" : "white";
  game.castlingRights = parseCastling(parts[2] ?? "-");
  game.enPassantTarget = parseEnPassant(parts[3] ?? "-");
  game.halfmoveClock = parts[4] ? parseInt(parts[4], 10) : 0;
  game.fullmoveNumber = parts[5] ? parseInt(parts[5], 10) : 1;
  game.resetLoadedPosition();
  return game;
}

function parseFenRank(rankStr: string, row: number, board: (Piece | null)[][]): void {
  let col = 0;
  for (const ch of rankStr) {
    if (col >= BOARD_SIZE) break;
    if (ch === ".") {
      board[row][col++] = null;
      continue;
    }
    if (ch >= "1" && ch <= "9") {
      const empty = parseInt(ch, 10);
      for (let i = 0; i < empty && col < BOARD_SIZE; i++) {
        board[row][col++] = null;
      }
      continue;
    }
    board[row][col++] = parsePieceChar(ch);
  }
  while (col < BOARD_SIZE) {
    board[row][col++] = null;
  }
}

function parsePieceChar(ch: string): Piece {
  const kind = ch.toUpperCase() as PieceKind;
  return { kind, color: ch === ch.toUpperCase() ? "white" : "black" };
}

function parseCastling(text: string): CastlingRights {
  if (text === "-") {
    return {
      whiteKingSide: false,
      whiteQueenSide: false,
      blackKingSide: false,
      blackQueenSide: false,
    };
  }
  return {
    whiteKingSide: text.includes("K"),
    whiteQueenSide: text.includes("Q"),
    blackKingSide: text.includes("k"),
    blackQueenSide: text.includes("q"),
  };
}

function parseEnPassant(text: string): Square | null {
  if (text === "-") return null;
  return fromEngineNotation(text) ?? fromStandardNotation(text);
}

function fenRank(game: ChessGame, row: number): string {
  let result = "";
  let empty = 0;
  for (let col = 0; col < BOARD_SIZE; col++) {
    const piece = game.board[row][col];
    if (piece) {
      if (empty > 0) {
        result += String(empty);
        empty = 0;
      }
      const ch = piece.kind;
      result += piece.color === "white" ? ch : ch.toLowerCase();
    } else {
      empty += 1;
    }
  }
  // Run-length "10" is invalid for API/engine FEN (digit 0); use dots like variants.ini.
  if (empty > 0) {
    result += empty >= BOARD_SIZE ? ".".repeat(empty) : String(empty);
  }
  return result;
}

function fenCastling(game: ChessGame): string {
  let s = "";
  if (game.castlingRights.whiteKingSide) s += "K";
  if (game.castlingRights.whiteQueenSide) s += "Q";
  if (game.castlingRights.blackKingSide) s += "k";
  if (game.castlingRights.blackQueenSide) s += "q";
  return s || "-";
}

function matchLegalMove(
  game: ChessGame,
  from: Square,
  to: Square,
  promotion?: PieceKind
): ReturnType<ChessGame["legalMoves"]>[0] | null {
  const color = game.activeColor;
  const strict = game.legalMoves(color).filter(
    (m) =>
      squaresEqual(m.from, from) &&
      squaresEqual(m.to, to) &&
      (promotion === undefined || m.promotion === promotion)
  );
  if (strict[0]) return strict[0];
  return (
    game.legalMoves(color).find(
      (m) => squaresEqual(m.from, from) && squaresEqual(m.to, to)
    ) ?? null
  );
}

/** Match Fairy-Stockfish UCI to a legal move (mirrors Swift move(fromEngineUCI:)). */
export function matchEngineMove(
  game: ChessGame,
  uci: string
): ReturnType<ChessGame["legalMoves"]>[0] | null {
  for (const { from, to, promotion } of resolveUciInterpretations(uci)) {
    const matched = matchLegalMove(game, from, to, promotion);
    if (matched) return matched;
  }
  return null;
}

File: ChessBorder/ChessBorder/Engine/FEN.swift
import Foundation

extension ChessGame {
    /// FEN for Fairy-Stockfish (10×10 border chess, ranks 10→1 top to bottom).
    func toFEN() -> String {
        var ranks: [String] = []
        for row in 0..<BoardConstants.size {
            ranks.append(fenRank(row: row))
        }
        let placement = ranks.joined(separator: "/")
        let side = activeColor == .white ? "w" : "b"
        let castle = fenCastling()
        let ep = enPassantTarget?.engineNotation ?? "-"
        return "\(placement) \(side) \(castle) \(ep) \(halfmoveClock) \(fullmoveNumber)"
    }

    private func fenRank(row: Int) -> String {
        var result = ""
        var empty = 0
        for col in 0..<BoardConstants.size {
            if let piece = board[row][col] {
                if empty > 0 {
                    result += fenEmptyRun(empty)
                    empty = 0
                }
                result += fenCharacter(for: piece)
            } else {
                empty += 1
            }
        }
        if empty > 0 {
            result += fenEmptyRun(empty)
        }
        return result
    }

    /// Run-length `10` is invalid for API/engine FEN (digit 0); use dots like variants.ini.
    private func fenEmptyRun(_ count: Int) -> String {
        count >= BoardConstants.size
            ? String(repeating: ".", count: count)
            : String(count)
    }

    private func fenCharacter(for piece: Piece) -> String {
        let ch = String(piece.kind.rawValue)
        return piece.color == .white ? ch : ch.lowercased()
    }

    private func fenCastling() -> String {
        var s = ""
        if castlingRights.whiteKingSide { s += "K" }
        if castlingRights.whiteQueenSide { s += "Q" }
        if castlingRights.blackKingSide { s += "k" }
        if castlingRights.blackQueenSide { s += "q" }
        return s.isEmpty ? "-" : s
    }

    /// Parse standard UCI (e2e4) for human/bot moves on the inner 8×8.
    func move(from uci: String) -> Move? {
        matchMove(uci: uci, preferEngineCoordinates: false)
    }

    /// Parse Fairy-Stockfish UCI (10×10 files a–j, ranks 1–10).
    func move(fromEngineUCI uci: String) -> Move? {
        matchMove(uci: uci, preferEngineCoordinates: true)
    }

    private func matchMove(uci: String, preferEngineCoordinates: Bool) -> Move? {
        let trimmed = uci.trimmingCharacters(in: .whitespaces).lowercased()

        var candidates: [(Square, Square, PieceKind?)] = []

        if preferEngineCoordinates, let parsed = UCIParser.parseEngineMove(trimmed) {
            candidates.append((parsed.from, parsed.to, parsed.promotion))
        }

        if trimmed.count >= 4 {
            let fromStr = String(trimmed.prefix(2))
            let toStr = String(trimmed.dropFirst(2).prefix(2))
            let promoChar = trimmed.count > 4 ? trimmed.last! : nil
            let promotion = promoChar.flatMap { ch in
                PieceKind(rawValue: Character(String(ch).uppercased()))
            }
            if let from = Square.fromStandardNotation(fromStr), let to = Square.fromStandardNotation(toStr) {
                candidates.append((from, to, promotion))
            }
            let fromMix = Square.fromStandardNotation(fromStr) ?? Square.fromEngineNotation(fromStr)
            let toMix = Square.fromStandardNotation(toStr) ?? Square.fromEngineNotation(toStr)
            if let fromMix, let toMix {
                candidates.append((fromMix, toMix, promotion))
            }
        }

        var seen = Set<String>()
        for (from, to, promotion) in candidates {
            let key = "\(from.row),\(from.col),\(to.row),\(to.col),\(promotion?.rawValue ?? "")"
            if seen.contains(key) { continue }
            seen.insert(key)

            let strict = legalMoves(for: activeColor).filter {
                $0.from == from && $0.to == to && $0.promotion == promotion
            }
            if let move = strict.first { return move }
            if let move = legalMoves(for: activeColor).first(where: { $0.from == from && $0.to == to }) {
                return move
            }
        }
        return nil
    }
}
